torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1555,12 +1555,62 @@ void ProcessGroupNCCL::groupEnd() {
   --ncclActiveGroupCounter_;
 }
 
+#ifdef ENABLE_NCCL_P2P_SUPPORT
+std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::alltoall(
+    std::vector<at::Tensor>& outputTensors,
+    std::vector<at::Tensor>& inputTensors,
+    const AllToAllOptions& /* unused */) {
+  auto device = outputTensors[0].device();
+  for (size_t r = 0; r < outputTensors.size(); r++) {
+    check_gpu_single_tensor(outputTensors[r]);
+    check_gpu_single_tensor(inputTensors[r]);
+    TORCH_CHECK(device == outputTensors[r].device() && device == inputTensors[r].device(),
+      "Tensors must be on the same device")
+  }
+  std::vector<at::Tensor> inputTensor0 = {inputTensors[0]};
+  std::vector<at::Tensor> outputTensor0 = {outputTensors[0]};
+  return collective(
+    inputTensor0,
+    outputTensor0,
+    [&](at::Tensor& /* unused */,
+        at::Tensor& /* unused */,
+        ncclComm_t comm,
+        at::cuda::CUDAStream& stream) {
+      C10D_NCCL_CHECK(ncclGroupStart());
+      for (size_t r = 0; r < outputTensors.size(); r++) {
+        at::Tensor &input = inputTensors[r];
+        at::Tensor &output = outputTensors[r];
+        if (input.numel() != 0) {
+          C10D_NCCL_CHECK(ncclSend(
+              input.data_ptr(),
+              input.numel(),
+              getNcclDataType(input.scalar_type()),
+              r,
+              comm,
+              stream.stream()));
+        }
+        if (output.numel() != 0) {
+          C10D_NCCL_CHECK(ncclRecv(
+              output.data_ptr(),
+              output.numel(),
+              getNcclDataType(output.scalar_type()),
+              r,
+              comm,
+              stream.stream()));
+        }
+      }
+      C10D_NCCL_CHECK(ncclGroupEnd());
+      return ncclSuccess;
+    });
+}
+#else
 std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::alltoall(
     std::vector<at::Tensor>& /* unused */,
     std::vector<at::Tensor>& /* unused */,
     const AllToAllOptions& /* unused */) {
   throw std::runtime_error("ProcessGroupNCCL does not support alltoall");
```

Comment:
Once you move this into the ifdef ENABLE_NCCL_P2P_SUPPORT above, you should also change the else error message...  
"ProcessGroupNCCL only supports recv for NCCL lib version >= 2.7.0"

Reply:
done